compiler/rustc_mir_transform/src/simplify.rs

```diff
@@ -645,6 +673,22 @@ fn remove_unused_definitions_helper(used_locals: &mut UsedLocals, body: &mut Bod
     while modified {
         modified = false;
 
+        if !used_locals.preserve_debug {
+            body.var_debug_info.retain(|info| {
+                let keep = match info.value {
+                    VarDebugInfoContents::Place(place) => used_locals.is_used(place.local),
+                    _ => true,
+                };
+
+                if !keep {
+                    trace!("removing var_debug_info {:?}", info);
+                    modified = true;
+                }
```

Comment:
Changing `modified` here is not useful: we don't update `used_locals`, so we do not mark new locals as unused.